pkg/cmd/roachtest/option/cluster_options.go

```diff
@@ -0,0 +1,182 @@
+		if o.ConnectionOptions == nil {
+			o.ConnectionOptions = make(map[string]string)
+		}
+
+		o.ConnectionOptions[key] = value
+	}
+}
+
+// ConnectTimeout allows callers to set a connection timeout.
+func ConnectTimeout(t time.Duration) OptionFunc {
+	sec := int64(t.Seconds())
+	if sec < 1 {
+		sec = 1
+	}
+	return ConnectionOption("connect_timeout", fmt.Sprintf("%d", sec))
+}
+
+// DBName changes the database name used when connecting to crdb.
+func DBName(dbName string) OptionFunc {
+	return func(o *GlobalOptions) {
+		o.DBName = dbName
+	}
+}
+
+// AuthMode allows the callers to change the authentication mode used
+// when connecting to crdb.
+func AuthMode(authMode install.PGAuthMode) OptionFunc {
+	return func(o *GlobalOptions) {
+		o.AuthMode = authMode
+	}
+}
+
+// Apply takes in an options struct and a list of user-provided
+// options, and applies those options to the container. The container
+// should be a pointer to a struct containing the relevant fields --
+// in other words, the struct actually defines that custom options a
+// function can take. The struct is expected to have a field for each
+// custom option passed. An error is returned if an unrecognized
+// option is passed.
+func Apply(container any, opts ...OptionFunc) (retErr error) {
+	var globalOptions GlobalOptions
+	for _, opt := range opts {
+		opt(&globalOptions)
+	}
+
+	// Many functions in the `reflect` package can panic if they
+	// parameters are of unexpected types, so we wrap these panics as
+	// errors to be returned to the caller.
+	var currentField string
+	defer func() {
+		if r := recover(); r != nil {
+			if currentField == "" {
+				retErr = fmt.Errorf("option.Apply failed: %v", r)
+			} else {
+				retErr = fmt.Errorf("failed to set %q on %T: %v", currentField, container, r)
+			}
+		}
+	}()
+
+	// Build a mapping from option to name to the corresponding
+	// `reflect.Value`.
+	globalOptionsValue := reflect.ValueOf(globalOptions)
+	globalFields := make(map[string]reflect.Value)
+	for _, f := range reflect.VisibleFields(reflect.TypeOf(globalOptions)) {
+		globalFields[f.Name] = globalOptionsValue.FieldByName(f.Name)
+	}
+
+	// We keep a set of fields from `globalOptions` that are actually
+	// used by the container struct so that we can validate that the
+	// caller didn't pass any options that are not applicable.
+	containerFields := make(map[string]struct{})
+
+	containerStruct := reflect.ValueOf(container).Elem()
+	containerType := reflect.TypeOf(containerStruct.Interface())
+	for _, structField := range reflect.VisibleFields(containerType) {
+		currentField = structField.Name
+		f := containerStruct.FieldByName(currentField)
+
+		// It is an error for the container struct to have fields that are
+		// not present in the `GlobalOptions` struct.
+		globalField, ok := globalFields[currentField]
+		if !ok {
+			return fmt.Errorf("options struct %T has unknown option %q", container, currentField)
+		}
+
+		f.Set(globalField)
```

Comment:
Oof, this is really the `apply` part, the rest is validation. Maybe worth a comment to help guide the reader?

Reply:
Done.